pyls/_utils.py

```diff
@@ -9,21 +9,49 @@
 
 FIRST_CAP_RE = re.compile('(.)([A-Z][a-z]+)')
 ALL_CAP_RE = re.compile('([a-z0-9])([A-Z])')
+KWD_MARK = object()
 
 
-def debounce(interval_s):
+def debounce(interval_s, keys=None):
     """Debounce calls to this function until interval_s seconds have passed."""
     def wrapper(func):
+        arg_cache = {}
+        lock = threading.Lock()
+
+        @functools.wraps(func)
+        def cleanup(*args, **kwargs):
+            input_hash = _hash_input(keys, *args, **kwargs)
```

Comment:
Hoist computing the input_hash up to where you create the lock

Reply:
you can't do that! the input_hash depends on the input and has to be recomputed each time the function is called. meanwhile the lock is just a member of the debounce closure.

Reply:
Oh yup, you're right. You could pass the key to `cleanup` as an argument. But not sure you gain much